rfd/0174-kube-rbac-bootstrap.md

```diff
@@ -0,0 +1,346 @@
+To improve this situation we will introduce auto-provisioning of the Kubernetes RBAC based on 
+the Teleport roles, which means users will be able to define required permissions completely in the Teleport, and under the hood
+we will keep them in sync transparently for the users.
+
+We will add a new field to the role definition - `kubernetes_permissions`. That field will contain desired Kubernetes RBAC permissions
+for that role and will be similar in structure to the native Kubernetes RBAC:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-role
+spec:
+  allow:
+    kubernetes_labels:
+      "*": "*"
+    kubernetes_groups: ["kube_group"]
+    kubernetes_permissions:
+      namespaces: ["namespace1", "namespace2"]
+      rules:
+      - resources: ["pod", "pod/exec"]
+        verbs: ["get", "list", "create"]
+      - resources: ["deployments"]
+        apiGroups: ["app"]
+        verbs: ["*"]
+      - resources: ["secrets"]
+        apiGroups: [""]
+        resourceNames: ["secret1", "secret2"]
+        verbs: ["get", "list""]
+```
+
+Presence of this field will signal that user wants Teleport to auto-provision these permissions into the Kubernetes clusters.
+Teleport will translate permissions defined in that field into Kubernetes RBAC definitions and will automatically create roles and bindings
+on the target Kubernetes clusters (as specified by the labels). If `namespaces` field includes '*' Teleport will create ClusterRole/ClusterRoleBinding
+pair on the Kubernetes cluster, otherwise Role/Binding pair will be created in the each listed namespace. 
+
+Subject for the bindings will be a group named `teleport:*RoleName*`, e.g. for a Teleport role `kube-role` the resulting 
+group name will be `teleport:kube-role`. This will allow Kube service to setup correct impersonation group headers based on the roles available to the user.
+
+If `kubernetes_permissions` field is set in the `allow` part of the Teleport role, fields `kubernetes_resources`, `kubernetes_users` and `kubernetes_groups` 
+should not be set, we will return an error if user will try to create a role mixing those fields. In the future we might relax this rule.
+
+The field `kubernetes_permissions` will be forbidden from appearing in the `deny` part of the role definition since we are trying to keep its meaning close to
+native Kubernetes RBAC, which can only allow access. Although the field `kubernetes_resource` could appear in the `deny` part of the role, 
+even if `kubernetes_permissions` is set in the `allow` part, giving user more flexibility using Teleport-level permissions check.
+
+Pattern matching for resource names that is supported by the `kubernetes_resources` field will not be supported (same as in native Kubernetes RBAC)
+by the `kubernetes_permissions` in the
+first implementation, but we will be able to add it later. Not allowing pattern matching, though limiting functionality compared to the `kubernetes_resources` field,
```

Comment:
I don't believe we can implement this correctly and safely. First, it would require polling all resources, matching them against the keyword, and updating the role to allow access to a subset. This means we would need to process everything in real time to avoid leaking resources to users, and we would likely encounter role name filter limits.

Given this, I think we will always need a combination of `kubernetes_permissions` and `kubernetes_resources` because they function differently. One provisions Kubernetes RBAC, while the other dynamically filters resources. If our intention is for `kubernetes_permissions` to become `kubernetes_resources` and enforce filters dynamically, I have concerns about its feasibility.

`kubernetes_permissions` and `kubernetes_resources` working together seems a better approach but I wonder if people will understand how to use them.

Reply:
Yes, I mentioned that we might need additional checks relying on `kubernetes_resources`. What I had in mind that we could translate something like `resourceNames: ['env-*`]` into `resourceNames: ["*"]` on the Kubernetes cluster role and add `kubernetes_resources` filter for `env-*`, so we get all resources from the Kubernetes cluster and filter them on Teleport. We could do that transparently for the user, so they don't have to set up `kubernetes_resources` manually.